### PR TITLE
[graphql] balance and coinConnection

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Implements balance and coinConnection. Deferring Balance.coinType and Coin.asMoveObject until we build out Move types. Update types that implement ObjectOwner to use these connections.

## Test Plan 

update snapshot, manual queries added to examples

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 1]
```diff
@@ -175,6 +182,110 @@ impl PgManager {
...
+        coin_type: Option<String>,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+    ) -> Result<Option<(Vec<StoredObject>, bool)>, Error> {
+        let mut query = objects::dsl::objects.into_boxed();
+        query = query
+            .filter(objects::dsl::owner_id.eq(address))
+            .filter(objects::dsl::owner_type.eq(1));
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:197]: Would it be useful to a comment to make it easier to remember what owner_type 1 is?

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 2]
```diff
@@ -175,6 +182,110 @@ impl PgManager {
...
+                .filter(objects::dsl::checkpoint_sequence_number.gt(after))
+                .order(objects::dsl::checkpoint_sequence_number.asc());
+        } else if let Some(before) = before {
+            let before = self.parse_obj_cursor(&before)?;
+            query = query
+                .filter(objects::dsl::checkpoint_sequence_number.lt(before))
+                .order(objects::dsl::checkpoint_sequence_number.desc());
+        }
+
+        let limit = first.or(last).unwrap_or(10) as i64;
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:215]: is 10 the default number we agreed upon? Is it somewhere documented?

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 1 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:197]: It's here so we can leverage the index on objects table, I had considered adding a comment but not sure if we should or not

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 3]
```diff
@@ -751,6 +862,102 @@ impl PgManager {
...
         }
     }
+
+    pub(crate) async fn fetch_balance(
+        &self,
+        address: SuiAddress,
+        coin_type: String,
+    ) -> Result<Option<Balance>, Error> {
+        let address = address.into_vec();
+
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:872]: nit: remove empty row

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 2 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:215]: Jsonrpc defaults to 50, idk why I picked 10. We can have this as a method on PgManager and return 50

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 4]
```diff
@@ -751,6 +862,102 @@ impl PgManager {
...
+    pub(crate) async fn fetch_balances(
+        &self,
+        address: SuiAddress,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+    ) -> Result<Option<Connection<String, Balance>>, Error> {
+        let address = address.into_vec();
+
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:897]: nit spacing

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs (hunk 1 above)
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:197]: ok, I leave it at your discretion!

File: crates/sui-graphql-rpc/src/context_data/db_data_provider.rs [hunk 5]
```diff
@@ -948,3 +1155,17 @@ impl From<StoredEpochInfo> for Epoch {
         }
     }
 }
+
+impl TryFrom<StoredObject> for Coin {
+    type Error = Error;
+
+    fn try_from(o: StoredObject) -> Result<Self, Self::Error> {
+        let object_id = SuiAddress::try_from(o.object_id).map_err(|_| {
+            Error::Internal("Failed to convert object_id to SuiAddress".to_string())
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/db_data_provider.rs:1164]: might be missing {}

**Review by stefan-mysten — APPROVED** [review]

**Pull request merged**